### PR TITLE
fix(DB/Core): Dragonmaw Peon AI

### LINKED ISSUE #1707: Creatures in Shadowmoon Valley
Creatures in Shadowmoon Valley

##### SMALL DESCRIPTION:

The creatures doesn't do damage.

##### EXPECTED BLIZZLIKE BEHAVIOUR:

They should attack the players and do damage with their abilities
[NPC Dragonmaw Peon](https://www.wowhead.com/npc=22252/dragonmaw-peon#abilities)
1. [Sunder Armor](https://www.wowhead.com/spell=15572/sunder-armor)
2. [Kick](https://www.wowhead.com/spell=34802/kick)

##### CURRENT BEHAVIOUR:
 In Shadowmoon Valley, preciously in Netherwing Ledge, there are a lot NPC's Dragonmaw Poen. If you walk near them, they will enter in combat, but they just follow player, doesn't do any damage or something

##### STEPS TO REPRODUCE THE PROBLEM:

1. .go creature 78000
2. Enter in combat
3. Check will they attack, use their abilities & damage you

##### BRANCH(ES):

master

##### AC HASH/COMMIT:
 97e7e22

##### OPERATING SYSTEM:
 Windows

##### MODULES:
 NO

##### OTHER CUSTOMIZATIONS:
 NO

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/72642652-creatures-in-shadowmoon-valley?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Stoabrogga**
##### CHANGES PROPOSED:
This PR contains basically the changes from #1676 and #1678 with the following modifications:
- grant the kill credit before the "Vomit" (43327) spell is applied, otherwise the player does not get the quest credit (I think this was what pklloveyou meant in his review of PR #1676)
- added another suggestion from pklloveyou to force the peons facing the mutton before eating and despawning the mutton after finishing eating (looks better in my opinion)
- leave the variable "Tapped" in the code (was removed in the other PRs); it is used to avoid an already vomiting peon to run to a new mutton, which would otherwise look strange
- add all broadcast text IDs to the creature text SQL script
- add an additional creature text for the poisoned peon

###### ISSUES ADDRESSED:
Closes #1707 

##### TESTS PERFORMED:
tested build (Ubuntu 16.04, clang 7) and in-game

##### HOW TO TEST THE CHANGES:
This is a little bit complicated. It is not enough to just add the quest "A Slow Death" (11020), as @comix1988 correctly stated in #1676. Instead to prepare for this as GM you have to add the whole quest chain leading to the daily quest 11020. Please test as follows (with GM mode off!):

- remove all associated quests of the quest chain and get hated with the netherwing:
```
.quest remove 10870
.quest remove 11012
.quest remove 11013
.quest remove 11014
.quest remove 11019
.quest remove 11020
.modify reputation 1015 hated
.go -4862.29 410.73 61.0058 530
```
- kill the Dragonmaw Peons and check that they are using their new abilities "Kick" and "Sunder Armor"
- add and reward all associated quests of the quest chain:
```
.quest add 10870
.quest reward 10870
.quest add 11012
.quest reward 11012
.quest add 11013
.quest reward 11013
.quest add 11014
.quest reward 11014
.quest add 11019
.quest reward 11019
.quest add 11020
.additem 32502 12
```
- proceed with the quest "A Slow Death"

##### KNOWN ISSUES AND TODO LIST:
none

##### Target branch(es):
Master

### CODE REVIEW

File: src/server/scripts/Outland/zone_shadowmoon_valley.cpp [hunk 1]
```diff
@@ -556,6 +556,24 @@ class npc_enslaved_netherwing_drake : public CreatureScript
 /*#####
 # npc_dragonmaw_peon
 #####*/
+enum DragonmawPeon
```
Comment (Pondaveia) [inline src/server/scripts/Outland/zone_shadowmoon_valley.cpp]: codestyle: align all the equals for readability

File: src/server/scripts/Outland/zone_shadowmoon_valley.cpp [hunk 2]
```diff
@@ -571,64 +589,112 @@ class npc_dragonmaw_peon : public CreatureScript
     {
         npc_dragonmaw_peonAI(Creature* creature) : ScriptedAI(creature) { }
 
+        EventMap events;
```
Comment (Pondaveia) [inline src/server/scripts/Outland/zone_shadowmoon_valley.cpp]: codestyle: EventMap events, events2;

File: src/server/scripts/Outland/zone_shadowmoon_valley.cpp [hunk 3]
```diff
@@ -571,64 +589,112 @@ class npc_dragonmaw_peon : public CreatureScript
...
         void Reset()
         {
+            events.Reset();
+            events2.Reset();
             PlayerGUID = 0;
             Tapped = false;
-            PoisonTimer = 0;
+        }
+
+        void EnterCombat(Unit* /*who*/)
```
Comment (Pondaveia) [inline src/server/scripts/Outland/zone_shadowmoon_valley.cpp]: Don't forget to override functions :)

File: src/server/scripts/Outland/zone_shadowmoon_valley.cpp [hunk 4]
```diff
@@ -571,64 +589,112 @@ class npc_dragonmaw_peon : public CreatureScript
...
 
-            if (id)
+        void CreditPlayer()
+        {
+            if (PlayerGUID)
             {
-                me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_ONESHOT_EAT);
-                PoisonTimer = 15000;
+                Player* player = ObjectAccessor::GetPlayer(*me, PlayerGUID);
+                if (player && player->GetQuestStatus(11020) == QUEST_STATUS_INCOMPLETE)
```
Comment (Pondaveia) [inline src/server/scripts/Outland/zone_shadowmoon_valley.cpp]: Maybe create a constant for these magic numbers. Also the one line below.

File: src/server/scripts/Outland/zone_shadowmoon_valley.cpp [hunk 5]
```diff
@@ -571,64 +589,112 @@ class npc_dragonmaw_peon : public CreatureScript
...
+                    player->KilledMonsterCredit(23209, 0);
             }
         }
 
         void UpdateAI(uint32 diff)
         {
-            if (PoisonTimer)
+            events.Update(diff);
```
Comment (Pondaveia) [inline src/server/scripts/Outland/zone_shadowmoon_valley.cpp]: Just for notice, you can have the same eventmap for both Out and in combat. As long as the checks out of combat are inside !updatevictim() they will never be triggered. So you can get rid of 1 of the events variable.

File: src/server/scripts/Outland/zone_shadowmoon_valley.cpp [hunk 6]
```diff
@@ -571,64 +589,112 @@ class npc_dragonmaw_peon : public CreatureScript
...
-                    PoisonTimer = 0;
+
+                case EVENT_WALK_TO_MUTTON:
+                    me->SetWalk(true);
+                    me->GetMotionMaster()->MovePoint(1, x, y, z, true);
+                    me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_STATE_NONE);
+                    me->HandleEmoteCommand(EMOTE_ONESHOT_CHEER);
+                    break;
+                case EVENT_POISONED:
+                    if (GameObject* food = me->FindNearestGameObject(185893, 5.0f))
```
Comment (Pondaveia) [inline src/server/scripts/Outland/zone_shadowmoon_valley.cpp]: Constant for magic numbers

File: data/sql/updates/pending_db_world/rev_1555327481238291012.sql [hunk 7]
```diff
@@ -0,0 +1,11 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1555327481238291012');
+
+DELETE FROM `creature_text` WHERE `creatureid` = '22252';
```
Comment (Meltie2013) [inline data/sql/updates/pending_db_world/rev_1555327481238291012.sql]: Creature Id doesn't require ' ' for the entries, but can be still used.

**Comment by Stoabrogga** [comment]
@Pondaveia @Meltie2013 
Thanks for your feedback. I updated the PR accordingly and also added an additional creature text for the poisoned peon which was missing.

**Comment by Pondaveia** [comment]
Tested and only found this issue:
When you put the food down, the peons will attack the player when they get into aggro range

**Comment by Stoabrogga** [comment]
@Pondaveia 
It works for me, they are not attacking me when they are friendly. Are you sure you included all steps while testing? I tested like this:

1. Reset quest chain and reputation with the Netherwing, teleport to Netherwing Ledge:
```
.quest remove 10870
.quest remove 11012
.quest remove 11013
.quest remove 11014
.quest remove 11019
.quest remove 11020
.modify reputation 1015 hated
.go -4862.29 410.73 61.0058 530
```
As you don't went through the quest chain all Dragonmaw Peons are hostile and will attack on sight (this is for testing if they are using the correct abilities).

2. Get all necessary quests of the quest chain up to the daily quest "A Slow Death" and provide the item "Fel Gland":
```
.quest add 10870
.quest reward 10870
.quest add 11012
.quest reward 11012
.quest add 11013
.quest reward 11013
.quest add 11014
.quest reward 11014
.quest add 11019
.quest reward 11019
.quest add 11020
.additem 32502 12
```
You automatically get the Fel Orc disguise aura and all Dragonmaw Peons are now friendly. Place the mutton before them and they will just eat it and die. That's at least what happens for me. They are not attacking.

**Comment by Stoabrogga** [comment]
Here's a small video showing how it looks on my server (only part 2: poisoning the peons):
[dragonmaw_peon.zip](https://github.com/azerothcore/azerothcore-wotlk/files/3083324/dragonmaw_peon.zip)

**Comment by Pondaveia** [comment]
I thought i wasn't supposed to have the aura and i removed it, in this case i think the PR is good

**Pull request merged**